### PR TITLE
Server side redirect on main page is added.

### PR DESCRIPTION
**Pull request by ANtlord**
## Contributor Checklist:

* [ ] I have updated the unit tests
* [ ] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

**Comment by ANtlord** [comment]
Please consider new relations. I'm not sure that it's good place for class `InternalResource`. Hope you have @tardyp an opinion about that. Thanks

File: master/buildbot/www/resource.py [hunk 1]
```diff
@@ -117,3 +119,19 @@ def render(self, request):
         redir = self.base_url + self.basepath
         request.redirect(redir)
         return redir
+
+
+class InternalResource(Resource):
+    """
+    Resource available only for authorized users if authorization is enabled.
+    """
+    def render(self, request):
```
Comment (tardyp) [inline master/buildbot/www/resource.py]: I don't think it is needed to factorize an InternalResource if this is not needed. (principle of Minimum Viable Product)

File: master/buildbot/www/resource.py [hunk 2]
```diff
@@ -117,3 +119,19 @@ def render(self, request):
...
+        from .oauth2 import OAuth2Auth
```
Comment (tardyp) [inline master/buildbot/www/resource.py]: avoid this kind of import

File: master/buildbot/www/resource.py [hunk 3]
```diff
@@ -117,3 +119,19 @@ def render(self, request):
...
+        wwwconfig = self.master.www
+        userinfos = wwwconfig.getUserInfos(request)
+        auth = wwwconfig.auth
+        if 'anonymous' in userinfos and isinstance(auth, OAuth2Auth) and auth.autologin:
```
Comment (tardyp) [inline master/buildbot/www/resource.py]: I would rather say:
```
 if 'anonymous' in userinfos and auth and auth.autologin:
     return auth.do_autologin(self, request)
```
and then implement the details of the autologin in the auth plugin.

File: master/buildbot/www/resource.py (hunk 1 above)
Comment (tardyp) [reply master/buildbot/www/resource.py]: so you can just implment this in the render function of IndexResource

**Comment by tardyp** [comment]
any luck on finishing this? 90% of the PR not finished after 3 days never finish :)

**Comment by ANtlord** [comment]
@tardyp I'm of the those who can't sleep with unfinished deals

**Comment by tardyp** [comment]
looks like the tests shows some corner cases need to be taken care of.
We also need a new test for this, and a newsfragment. 
http://docs.buildbot.net/1.4.0/developer/quickstart.html for how to run tests locally

**Comment by ANtlord** [comment]
Sorry for the delay. I don't forget about this PR. I'll take a time for it ASAP.

**Comment by ANtlord** [comment]
@tardyp There is an error related to condition checks anonymous and check `auth` object. The text the error has is `exceptions.AttributeError: 'NoAuth' object has no attribute 'autologin'`. It would be avoided If I remain my check of class of an object. If it shouldn't be in the code still I have no idea how to prevent the error. Here is a link to the build https://nine.buildbot.net/#/builders/8/builds/2866. Checkout it please. Any idea?

**Comment by tardyp** [comment]
@ANtlord I would rather add the attribute autologin=False by default in the BaseAuth class.

**Comment by ANtlord** [comment]
> @ANtlord I would rather add the attribute autologin=False by default in the BaseAuth class.

Are you sure? I can but it would change the entire interface

**Comment by tardyp** [comment]
> @ANtlord I would rather add the attribute autologin=False by default in the BaseAuth class.

> Are you sure? I can but it would change the entire interface

yes, by modifying config.py you are already changing the interface.

**Comment by ANtlord** [comment]
@tardyp Hello! Would you mind to look at the [output](https://nine.buildbot.net/#/builders/10/builds/3036/steps/7/logs/stdio)? I'm not sure how could I affect that.

**Comment by ANtlord** [comment]
@tardyp It doesn't do. `bb` build fails and I don't understand why, do you? I've attached link to the error in the previous comment. Look at there please.

**Comment by p12tic** [comment]
Sorry for chiming in this late into the PR. Would a new test make sense? There are similar tests in test/unit/test_www_oauth.py, shouldn't be hard to add..

**Comment by p12tic** [comment]
@ANtlord The failure was caused by old Chrome being installed on the test images. This shouldn't be a problem anymore on latest master.

**Comment by p12tic** [comment]
Also, can we have at least a little documentation and a release note? I would have added the unit test myself, but I was unsure what exact problem this PR is solving.
